lua/ttt2/libraries/outline.lua

```diff
@@ -20,37 +25,172 @@ OUTLINE_MODE_BOTH = 0
 local COLOR = 2
 local MODE = 3
 
+--FIXME: You can decide if you want this to be a option for users, should be adressed before merge
+local enable_thin_line_workaround = false
+
+local function SetRenderType(render_type)
+    if (
+        render_type ~= OUTLINE_RENDERTYPE_BEFORE_VM and
+        render_type ~= OUTLINE_RENDERTYPE_BEFORE_EF and
+        render_type ~= OUTLINE_RENDERTYPE_AFTER_EF
+    ) then
+        return
+    end
+
+    RenderType = render_type
+
+    return true
+end
+
+local function GetRenderType()
+    return RenderType
+end
+
+
+local function GetOutlineThickness()
+    return OutlineThickness
+end
+
+local function SetOutlineThickness(thickness)
+    if thickness < 1 then
+        return
+    end
+
+    OutlineThickness = thickness
+
+    return true
+end
+
+
+local function GetCurrentLists()
+    return Lists[GetRenderType()]
+end
+
+local function GetCurrentListsSize()
+    return ListsSize[GetRenderType()]
+end
+
+
+local function GetCurrentList()
+    return Lists[GetRenderType()][GetOutlineThickness()]
+end
+
+local function GetCurrentListSize()
+    return ListsSize[GetRenderType()][GetOutlineThickness()]
+end
+
+
+local function ResetCurrentLists()
+    Lists[GetRenderType()] = {}
+end
+
+local function ResetCurrentListsSize()
+    ListsSize[GetRenderType()] = {}
+end
+
+
+local function SetCurrentList(value)
+    Lists[GetRenderType()] = value
+end
+
+local function SetCurrentListSize(size)
+    ListsSize[GetRenderType()][GetOutlineThickness()] = size
+end
+
+
+local function InitializeCurrentListIfNeeded()
+    local render_type, outline_thickness = GetRenderType(), GetOutlineThickness()
+
+    if not Lists[render_type] then
+        Lists[render_type] = {}
+    end
+
+    if not ListsSize[render_type] then
+        ListsSize[render_type] = {}
+    end
+
+    if not Lists[render_type][outline_thickness] then
+        Lists[render_type][outline_thickness] = {}
+    end
+
+    if not ListsSize[render_type][outline_thickness] then
+        ListsSize[render_type][outline_thickness] = 0
+    end
+end
+
 ---
 -- @param table ents List of @{Entity}
 -- @param Color color
 -- @param number mode [OUTLINE_MODE_BOTH, OUTLINE_MODE_NOTVISIBLE, OUTLINE_MODE_VISIBLE]
 -- @realm client
-function outline.Add(ents, color, mode)
+function outline.Add(ents, color, mode, render_type, outline_thickness)
+    -- Make the default behaviour comply with older revisions
+    if not render_type then
+        render_type = OUTLINE_RENDERTYPE_AFTER_EF
+    end
+
+    if not outline_thickness then
+        outline_thickness = 1
+    end
+
+    -- Check for a validity
+    if not SetRenderType(render_type) then
+        return
+    end
+
+    if not SetOutlineThickness(outline_thickness) then
+        return
+    end
+
+    -- Create list if it doesnt exist
+    InitializeCurrentListIfNeeded()
+
+    -- Get List and ListSize for this render type and outline thickness
+    local List, ListSize = GetCurrentList(), GetCurrentListSize()
```

Comment:
variables in our lua style are always lower case - or is here a reason for this decision?

Reply:
because of https://wiki.facepunch.com/gmod/list
though we could find another name

Reply:
Ah, I see. I'd prefer a different name then. Something like `entsList`, `renderTargetList` or so maybe?

Reply:
Ah i remember why i did that, theres also a global Lists and ListSize, i used shadowing locals there to make development easier keeping it portable between different parts of the module, but it does break the no uppercase for non globals rule, can adjust that if you still want it